### PR TITLE
feat(SpoonifierVisitor): a visitor that generate the spoon code to recreate a spoon model

### PR DESCRIPTION
**Pull request by nharrand**
…red to recreate a spoon model

### CODE REVIEW

File: src/test/java/spoon/test/spoonifier/SpoonifierTest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+		testSpoonifierWith("src/test/java/spoon/test/prettyprinter/testclasses/FooCasper.java", i++);
+		testSpoonifierWith("src/test/java/spoon/test/prettyprinter/testclasses/Rule.java", i++);
+
+	}
+
+	/*
+	 * @pathToClass path to the class that will be Spoonified
+	 * @i an integer to avoid duplicated classes
+	 *
+	 * This method verify that SpoonifyVisitor can generate code that replicate a class
```
Comment (seintur) [inline src/test/java/spoon/test/spoonifier/SpoonifierTest.java]: verifies
replicates

Comment (nharrand) [reply src/test/java/spoon/test/spoonifier/SpoonifierTest.java]: Thanks

**Comment by nharrand** [comment]
There are still some issues.
Works on `src/test/java/spoon/test/spoonifier/testclasses/ArrayRealVector.java` but not on `src/test/java/spoon/test/prettyprinter/testclasses/ArrayRealVector.java`

**Comment by monperrus** [comment]
Really cool. What about adding an assertEquals(code, spoonify(elem)) so that we have a feeling of
the generated code?

**Comment by nharrand** [comment]
@monperrus Do you mean an assertion on the element itself like in src/test/java/spoon/test/spoonifier/SpoonifierTest.java:81 ? or do you mean an assertion on the code pretty printed?

**Comment by monperrus** [comment]
I mean an assertion on the code pretty printed.

**Comment by nharrand** [comment]
I was about to say that assertEquals on elements should be stronger then on the prettyprinted code, and that obviously two elements that are equal should be printed identically by construction. But of course I have tried...

**Comment by nharrand** [comment]
Ok it does not necessarily leads to the same pretty printing. If I understand correctly, one of the issue is containers that are map. Depending on the order of creation of object some elements contained might not be in the same order. (But the order does not carry semantic in this case.)
Example:
```java
@spoon.test.annotation.testclasses.AnnotParamTypes(
integer = 42, 
integers = { 42 }, 
string = "Hello World!", 
strings = { "Hello", "World" }, 
clazz = java.lang.Integer.class, 
classes = { java.lang.Integer.class, java.lang.String.class }, 
b = true, 
byt = 42, 
c = 'c', 
s = ((short) (42)), 
l = 42, 
f = 3.14F, 
d = 3.14159, 
e = AnnotParamTypeEnum.G, 
ia = @InnerAnnot("dd")
)
```
becomes:
```java
@spoon.test.annotation.testclasses.AnnotParamTypes(
strings = { "Hello", "World" }, 
string = "Hello World!", 
s = ((short) (42)), 
l = 42, 
integers = { 42 }, 
integer = 42, 
ia = @InnerAnnot("dd"), 
f = 3.14F, 
e = AnnotParamTypeEnum.G, 
d = 3.14159, 
clazz = java.lang.Integer.class, 
classes = { java.lang.Integer.class, java.lang.String.class }, 
c = 'c', byt = 42, 
b = true
)
```
I am not sure if we want to fix this, as I am not sure it is a problem. WDYT?

**Comment by monperrus** [comment]
> Some elements contained might not be in the same order. 

yes,

But I meant the generated code sorry!

**Comment by nharrand** [comment]
Ok I have added one.

File: src/main/java/spoon/experimental/SpoonifyVisitor.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+import spoon.reflect.visitor.CtScanner;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+public class SpoonifyVisitor extends CtScanner {
```
Comment (monperrus) [inline src/main/java/spoon/experimental/SpoonifyVisitor.java]: javadoc? + some words in the `docs` folder

File: src/test/java/spoon/test/spoonifier/SpoonifierTest.java [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+		//Invoke the code generated
+		URLClassLoader urlClassLoader = new URLClassLoader(new URL[]{outputBinDir.toURI().toURL()});
+		Class rtWrapper = urlClassLoader.loadClass("SpoonifierWrapper" + i);
+		Method get = rtWrapper.getMethod("get", Factory.class);
+		CtElement generatedElement = (CtElement) get.invoke(null, targetElement.getFactory());
+
+		//contract: The element created by the code generated by SpoonifyVisitor is equivalent to the visited element
+		assertEquals(targetElement, generatedElement);
+
+		//contract: The original element and the sponified element leads to the same code once pretty printed
```
Comment (monperrus) [inline src/test/java/spoon/test/spoonifier/SpoonifierTest.java]: remove

**Comment by monperrus** [comment]
LGTM, will merge.

**Comment by nharrand** [comment]
@monperrus CI is finally happy!

**Comment by monperrus** [comment]
Cool, thanks!

**Pull request merged**